### PR TITLE
changes for better memory usage in users of BottomUp (parallel_rmr, etc)

### PR DESCRIPTION
**Pull request by bertschinger**
The first two commits are changes that reduce the size of `struct BottomUp`:

- removes a mostly unused embedded `struct stat` 
- removes two fixed 4 KiB buffers, one of which is always unused for parallel_rmr, in favor of dynamically allocating a buffer for pathnames that is only as large as needed

Overall this reduces the size of `struct BottomUp` from 8504 B to 184 B and seems to make a significant difference for parallel_rmr.

The third commit in the series modifies the logic of `descend_to_bottom()` in BottomUp.c to not format pathnames for non-directory nodes unless those nodes are actually processed. This removes an allocation for all the non-dir nodes in cases where those nodes are not processed.

### CODE REVIEW

File: src/BottomUp.c [hunk 1]
```diff
@@ -240,48 +343,51 @@ static int descend_to_bottom(QPTPool_t *ctx, const size_t id, void *data, void *
...
+                continue;
+            }
 
-            new_work.alt_name_len += sqlite_uri_path(new_work.alt_name + new_work.alt_name_len,
-                                                     sizeof(new_work.alt_name) - new_work.alt_name_len,
-                                                     entry->d_name, &name_len);
+            if (S_ISDIR(st.st_mode))
+                is_dir = 1;
+        } else if (entry->d_type == DT_DIR) {
+            is_dir = 1;
```
Comment (calccrypto) [inline src/BottomUp.c:360]: This can probably be merged into 1 if statement

File: src/utils.c [hunk 2]
```diff
@@ -1051,3 +1042,24 @@ ssize_t read_size(const int fd, void *buf, const size_t size) {
...
+    int d_fd = dirfd(d);
+    if (d_fd < 0) {
+        /*
+         * We should never get here. glibc's dirfd(3) never return errors, and
+         * Apple's libc only returns an error if the DIR * is invalid, which would
+         * indicate a bug in GUFI.
+         */
+        fprintf(stderr, "BUG: dirfd(3) failed: errno = %d\n", errno);
+        exit(1);
+    }
```
Comment (calccrypto) [inline src/utils.c:1062]: Leave the comment and remove the check?

File: src/BottomUp.c [hunk 3]
```diff
@@ -415,17 +521,16 @@ int parallel_bottomup_enqueue(QPTPool_t *pool,
     struct UserArgs *ua = NULL;
     QPTPool_get_args(pool, (void **) &ua);
 
-    struct BottomUp *root = (struct BottomUp *) malloc(ua->user_struct_size);
-    root->name_len = SNFORMAT_S(root->name, MAXPATH, 1, path, len);
+    struct BottomUp *root = (struct BottomUp *) calloc(ua->user_struct_size, 1);
+    new_pathname(root, path, len, NULL, 0);
 
     if (ua->generate_alt_name) {
         size_t name_len = root->name_len;
```
Comment (calccrypto) [inline src/BottomUp.c:528]: Don't need this variable any more

File: src/BottomUp.c [hunk 4]
```diff
@@ -108,6 +122,88 @@ struct UserArgs {
...
+ *   Fills in the path and path length in the given struct BottomUp.
+ */
+static void new_pathname(struct BottomUp *work, const char *dirname, size_t dirname_len,
+                         const char *basename, size_t basename_len) {
+    size_t new_len = 0; // does NOT count null terminator
+
+    if (basename)
+        new_len = dirname_len + basename_len + 1; // +1 for path separator
+    else
+        new_len = dirname_len;
```
Comment (calccrypto) [inline src/BottomUp.c:143]: `{}` even for only 1 line

File: src/utils.c (hunk 2 above)
Comment (bertschinger) [reply src/utils.c:1062]: I could go either way on this. This check does seem kind of stupid in a way, but on the other hand being explicit about errors does seem like "good hygiene".

Probably the only realistic reason to hit this would be accidentally forgetting to check the return of `opendir()` and passing a NULL pointer to `dirfd()`. On Linux using glibc, this would immediately segfault, but on Mac, this would return EINVAL.

In that case, the program would shortly crash anyways as most likely `opendir()` will be followed by `readdir()` which will dereference the NULL `DIR *`.

So basically, if this error is ignored, the program is gonna crash either way. Its just whether it segfaults, or if it prints out an error message and then exits.

File: src/BottomUp.c (hunk 3 above)
Comment (bertschinger) [reply src/BottomUp.c]: I ended up handling this a different way and changed `new_alt_pathname()` to return an error when the modified pathname is too big

**Comment by bertschinger** [comment]
In addition to changes for the review comments, the latest revision also does 2 things:
- changes `new_alt_pathname()` to return an error when it can't format the new pathname because it exceeds the buffer size, rather than just silently truncating it
- changes `new_alt_pathname()` to use a stack-allocated temporary buffer. Since the buffer was fixed-size, there was no point in heap allocating it

File: src/BottomUp.c [hunk 5]
```diff
@@ -108,6 +118,99 @@ struct UserArgs {
...
+            // uh-oh ... didn't convert entire string
+            return -1;
+        }
+    }
+
+    char *path = calloc(new_len + 1, sizeof(char));
+    if (!path)
+        goto oom;
+
+    strncpy(path, buf, new_len);
```
Comment (calccrypto) [inline src/BottomUp.c:201]: I recommend using `memcpy` or `SNFORMAT_S` to not have to do NULL char checks

File: src/BottomUp.c [hunk 6]
```diff
@@ -108,6 +118,99 @@ struct UserArgs {
...
+    if (!path) {
+        fprintf(stderr, "%s(): out of memory", __func__);
+        /* No point in continuing if we run out of memory: */
+        exit(1);
+    }
+
+    if (basename)
+        SNFORMAT_S(path, new_len + 1, 3, dirname, dirname_len, "/", (size_t) 1, basename, basename_len);
+    else
+        strncpy(path, dirname, new_len + 1);
```
Comment (calccrypto) [inline src/BottomUp.c:150]: style: always use `{}` in if blocks, even single lines

File: src/BottomUp.c [hunk 7]
```diff
@@ -108,6 +118,99 @@ struct UserArgs {
...
+
+        if (converted_len < dirname_len) {
+            // uh-oh ... didn't convert entire string
+            return -1;
+        }
+    }
+
+    char *path = calloc(new_len + 1, sizeof(char));
+    if (!path)
+        goto oom;
```
Comment (calccrypto) [inline src/BottomUp.c:199]: style: always use `{}` in if blocks, even single lines

Comment (bertschinger) [reply src/BottomUp.c]: I just got rid of the `goto` altogether since now there's only one error path in this function.
